### ISSUE TITLE
Add ability to share elements with 'unique' URL

### ISSUE THREAD

**Issue by wutfan** [BookStackApp/BookStack#288]
**Desired Feature:**
It would be great to have a way to share single books/chapters/pages with a single party without requiring them to have an account. I realize it is possible to just make an element available to the public but this is not a great solution if I want to share content with multiple parties. Let‘s say I want to share book A with person X and book B with person Y. I don‘t think there is a way to do this without sharing books A **and** B with X **and** Y by making both books publicly available. 
What I am thinking about is an obscured link with a hash like Google Docs does it.
I haven't been using BookStack for too long, so maybe I'm missing a clever way to achieve this with the existing permission system?

**Comment by ssddanbrown** [comment]
@wutfan Thanks for this request. I agree this would be a great feature and there's not currently a way to do this.

**Comment by tomershvueli** [comment]
I would also love to see this functionality!

**Comment by yoyokko** [comment]
https://github.com/koios-sh/BookStack

Hijack the authenticate and bypass the url prefix '/books/', '/search/', '/uploads/images/', set the permission of one book to public, close global 'public access', and I have this feature with minimum code changes. 

Hope for the official implements with QRCode url.

**Comment by jrnp97** [comment]
Any update? :)

**Comment by sense-design** [comment]
This would be a great feature

**Comment by Jeremylepr** [comment]
This feature really need to be implemented.

**Comment by Shootify** [comment]
yes please...

**Comment by caydenmb** [comment]
Agreed, I would utilize this feature a lot.

**Comment by Dylan-Miles** [comment]
This feature would be very useful!

**Comment by aware2** [comment]
+1 for this feature ❤️

**Comment by onthebackof** [comment]
Would be great 👍

**Comment by mfx-jgoetzinger** [comment]
+1

**Comment by artworklv** [comment]
I totally agree, having the ability to share things using a unique URL would be an excellent feature to have!

**Comment by scaphandroid** [comment]
+1 !

**Comment by skarados** [comment]
Yes please! +1!

**Comment by chdcomputers** [comment]
I also agree and I would like to add the following ideas:

1. To have control over the obscured link. Probably by assigning a GUID or something else as a unique ID to each of them.
2. Add expiration date so after a specified date and time the link won't be available any more.
3. To have the ability to share the same thing with more than one obscured links.
4. Add an optional description for the obscured link which will be logged (audited) along with the visits to those links
5. The visits to these links should trigger a webhook

Cheers!

**Comment by lucas-strummer** [comment]
+1! Thank you!

**Comment by Limerick-gh** [comment]
+1
That would be missing piece to establish Bookstack within my department, since we don't want the hassle of username/password authentication with the end-users.

**Comment by DanMundy** [comment]
+1

**Comment by DerMilderJoghurt** [comment]
+1

**Comment by fegrue** [comment]
+1

**Comment by virtadpt** [comment]
This would be really helpful - kind of like the sharable links that Wallabag has.

**Comment by megaxorg** [comment]
Would be great

**Comment by prplk** [comment]
+1

**Comment by ssddanbrown** [comment]
Been trying to think this through so just jotting down my thoughts so they're not stuck in my head.

### Proposal

- Share menu on content (books, chapters, pages, probably not shelves) to create share links.
- Share links lead to their own routes/controllers/views.
  - Super minimal UI & feature-set, with just content and page/context navigation.
  - Safer than punching through auth an adding extra visibility/access control.
  - Book/Chapter could complicate routing a tad with navigation within
    - plus would ideally need to handle internal cross-reference links.
- Role permission needed for creating share link, disabled by default.
- Overview of created links in settings/admin area.
- Probably would need an indication/view of existing share links in entity permission controls view.
- Share link would be entry in DB
  - Linked to entity
  - Stored creator & created date details.
  - Actual share unique ID a random gen 32 char (URL safe) value
     - Keep thinking this should have some added level of protection but we'd really want this to be accessible/reversible (to show and reuse as active share links) and any extra level of protection seems redundant. 

##### Permission Complications

- Not confident on what permissions these links should have. There's lots of options/assumptions that could be considered here. But I'm not keen on complicating matters here for *potential* needs and scenarios.
  - I'm leaning toward full view access for simplicity (in implementation and understanding/UX)
    - But then there's extra risk of over exposure which we'd probably need significant warnings at appropriate levels (for the role permission, when creating a link, when managing permissions) which spreads quite a bit of UI mess.
- Permissions will complicate image & attachment access (which we'd really need to provide).

**Comment by bmoellerTS** [comment]
ich würde mich über diese Funktion auch sehr freuen

**Comment by Thoroslives** [comment]
This feature is a must for me, I often make small little docs to share but have to export them at the moment as there's no way to share the individual doc with only that person without allowing all public to view all docs (this is without creating an account etc)

**Comment by CarrnellTech** [comment]
+1

**Comment by cgct** [comment]
+1

**Comment by edersong** [comment]
It's the only feature that this awesome application is lacking of 😀

**Comment by LegendaryFire** [comment]
Just search this up and ran across this thread actually, definitely a +1 from me as well.

**Comment by vytas911** [comment]
would be much appreciated +1

**Comment by netAction** [comment]
+1

The discussed features are much more than the desired features. Google Docs does not add any code to the share URL. You simply share the regular URL of that document. Viewing a shared Google Docs page is like Bookstack without shelves and books overview page. The guest role should have an option to disable "recent activity" panel, "recent pages" panel and /books.

Pro: The URL of a page is always the same. Everyone has the same deep link. Sharing can be done with the already existing guest role functionality. External users can use links between all books and pages until they reach something that is not shared with public.

Con: The names of books can be guessed. If your project's documentation has the url /books/alice, you know there will be something at /books/bob and /books/cindy. Easy fix: Always add a code to book URLs in case they might be shared later.

**Comment by netAction** [comment]
A simplification of the proposal from @ssddanbrown:

> Share menu on content

Yes to the menu but it could just create a reading permission for the guest group

> Share links lead to their own routes/controllers/views

I don't like multiple URLs for the same content very much and don't think it's neccessary.
   
> Super minimal UI & feature-set, with just content and page/context navigation.

Sounds great and would make Bookstack usable as poor man's CMS. Like for documentations.

Stripping out the lists of recent changes, list of books etc. even hides books that the receiver of the URL should not see. The minimal UI differentiates the books from each other.

> Book/Chapter could complicate routing a tad with navigation within
> plus would ideally need to handle internal cross-reference links.

Another reason for keeping the existing URLs. The OP mentioned Google Docs and they keep the URL too. Editors could even link between books.
    
> Overview of created links in settings/admin area.

This feature is useful and would be versatile if it would not cover share links but all group permissions.

> Probably would need an indication/view of existing share links in entity permission controls view.

Yes! A share link feels like guest group access anyway. It could simply be guest access.

> Keep thinking this should have some added level of protection but we'd really want this to be accessible/reversible (to show and reuse as active share links) and any extra level of protection seems redundant.

As a security feature, a random code could be added to every new book/chapter/page path. Just a random string at the end like /books/holiday-plans/page/trip-df5g6df5g6

The security might be even a reason against dedicated share links. Because share links can't hide one page of a shared book. Group access rules can.

**Comment by yrosman** [comment]
+1

**Comment by Thoroslives** [comment]
Is there any progress on this feature request? I feel its a big one and its not received any response from the dev for a good long while

**Comment by matteosavio** [comment]
+1

**Comment by MundMBo** [comment]
+1

**Comment by Kofl** [comment]
+1

**Comment by ssddanbrown** [comment]
Just noting down a potential avenue for this, for when I return to it, before it's lost from memory:

- We could internally use our own ZIP export system.
- Either a conversion, or direct rendering from ZIP format.
- On "share" content is internally "exported" to ZIP, and tracked against the content for management.
- Shares could be tracked to a specific revision/time (Probably time since that works across heirachy).
- Main downside could be the duplication of content (especially images/attachments).
- If we rely on storage, that could complicate things in complex (distributed/s3) environments.

**Comment by ssddanbrown** [comment]
BookStack is now [managed on Codeberg](https://codeberg.org/bookstack/bookstack).

As part of this change, we're closing all existing GitHub issues and pull requests.
You'll instead find this issue on Codeberg here: 
https://codeberg.org/bookstack/bookstack/issues/288
  
Any further discussion for this will now be done on Codeberg.
Therefore this GitHub thread will be locked.
